applications: asset_tracker_v2: Use Location library (#8631)
nrfconnect/sdk-nrf: applications/asset_tracker_v2/doc/location_module.rst

A location request starts when the module receives an ``APP_EVT_DATA_GET`` event and the ``APP_DATA_LOCATION`` type is listed in the event's ``data_list`` member containing the data types that shall be sampled.

The default configuration for the location request is set as follows:

* GNSS set as the first priority method and cellular positioning as the second priority.
  Methods that should not be used can be configured by adding them into ``No Data List``.
  See :ref:`Real-time configurations <real_time_configs>` for more information.
* Timeout which can be configured as part of :ref:`Real-time configurations <real_time_configs>`.

The location module receives configuration updates as payload in the following two events:

* ``DATA_EVT_CONFIG_INIT`` - The event contains the value configured at build time or received from cloud and stored to flash.
  The event is distributed from the data module as part of the application initialization.
* ``DATA_EVT_CONFIG_READY`` - The event is received when the data module has processed an incoming message from cloud with a configuration update.

The module sends the following events based on the outcome of the location request:

> * ``LOCATION_MODULE_EVT_GNSS_DATA_READY``; GNSS position has been acquired

Author: Changed to semicolon without prepending space.

> * ``LOCATION_MODULE_EVT_NEIGHBOR_CELLS_DATA_READY``: neighbor cell measurements are ready to be sent to cloud for location resolution

Reviewer: How about obtained?

> * ``LOCATION_MODULE_EVT_NEIGHBOR_CELLS_DATA_READY``: neighbor cell measurements are ready to be sent to cloud for location resolution

Author: "Obtained" is fine. I've tried to use obtained/retrieved/requested etc. for location and neighbor cell measurements.
Do you want me to remove this part of the comment: `are ready to be sent to cloud for location resolution`?

* ``LOCATION_MODULE_EVT_DATA_NOT_READY``: Location request failed
* ``LOCATION_MODULE_EVT_TIMEOUT``: Timeout occurred
* ``LOCATION_MODULE_EVT_AGPS_NEEDED``: A-GPS request should be sent to cloud
* ``LOCATION_MODULE_EVT_PGPS_NEEDED``: P-GPS request should be sent to cloud

GNSS LNA configuration
======================

Different devices have different GNSS antenna and LNA setups depending on the antenna type (onboard or external).
The application uses the :ref:`lib_modem_antenna` library for configuring the LNA.
The library has Kconfig options to control the LNA using either the COEX0 or MAGPIO interface of the nRF9160.
See the library documentation for more details on how to configure the antenna and LNA.

GPS assistance data
===================

The location module receives requests for GPS assistance data from the :ref:`lib_location` library.
When the module receives an A-GPS request, it distributes it to the other modules as a ``LOCATION_MODULE_EVT_AGPS_NEEDED`` event that contains information about the type of assistance data needed.
Providing the requested A-GPS data may reduce the time it takes to acquire a GNSS fix.

Module internals
================

> The location module is a threadless module in terms of message processing.

Author: Good that you found this. I was supposed to ask whether we should have own thread for the location module message processing.
For GNSS module, we had a thread for processing GNSS events from the modem library.
For cloud, data, debug, modem and sensor modules we have a thread for processing their messages. Rest of the modules don't have own thread.

> The location module is a threadless module in terms of message processing.

Author: @simensrostad Any comments on this?

> The location module is a threadless module in terms of message processing.

Reviewer: As long as none of the location API calls blocks for a prolonged amount of time there shouldn't be any need to have thread.

> The location module is a threadless module in terms of message processing.

Author: Location API calls are rather quick because the library has so long lasting functionality that any time consuming information is coming later in the event callback. I'll modify this text.

The thread is used to offload data processing from the location callbacks coming from the :ref:`lib_location` library, and no time consuming tasks should be performed there.

All incoming events from other modules are handled in the context of the Application Event Manager callback, because they all complete fast enough to not require a separate thread.

Configuration options
*********************

.. _CONFIG_LOCATION_MODULE:

CONFIG_LOCATION_MODULE
   Enables the location module.

Module states
*************

The location module has an internal state machine with the following states:

  * ``STATE_INIT`` - The initial state of the module in which it awaits the modem to be initialized and receive the location related configuration.
  * ``STATE_RUNNING`` - The module has performed all required initialization and can respond to requests to start a location request. The running state has two sub-states:

    * ``SUB_STATE_SEARCH`` - A location request is ongoing.
    * ``SUB_STATE_IDLE`` - The module is idling and can respond to a request to start a location request.

>   * ``STATE_SHUTDOWN`` - The module has been shut down after receiving a request from the util module.

Reviewer 2: ```suggestion
  * ``STATE_SHUTDOWN`` - The module has been shut down after receiving a request from the utility module.
```

> State transitions take place based on input from other modules, such as the app module, data module, util module and modem module.

Reviewer 2: ```suggestion
State transitions take place based on input from other modules, such as the application module, data module, utility module and modem module.
```

* ``DATA_EVT_CONFIG_READY`` - The event is received when the data module has processed an incoming message from cloud with a configuration update.

When a position has been retrieved, the ``LOCATION_MODULE_EVT_DATA_READY`` event is sent from the module.
If the location request failed, ``LOCATION_MODULE_EVT_DATA_NOT_READY`` is sent.
If a timeout occurred, ``LOCATION_MODULE_EVT_TIMEOUT`` is sent.

GNSS LNA configuration
======================

Different devices have different GNSS antenna and LNA setups depending on the antenna type (onboard or external).
The application uses the :ref:`lib_modem_antenna` library for configuring the LNA.
The library has Kconfig options to control the LNA using either the COEX0 or MAGPIO interface of the nRF9160.
See the library documentation for more details on how to configure the antenna and LNA.

GPS assistance data
===================

The location module receives requests for GPS assistance data from the :ref:`lib_location` library.
When the module receives an A-GPS request, it distributes it to the other modules as a ``LOCATION_MODULE_EVT_AGPS_NEEDED`` event that contains information about the type of assistance data needed.

> Providing the requested A-GPS data may reduce the time it takes to acquire a GNSS fix.

Author: Changed to:
`Providing the requested A-GPS data typically reduces the time it takes to acquire a GNSS fix significantly`

.. _asset_tracker_v2_location_module:

Location module
###############

.. contents::
   :local:
   :depth: 2

The location module controls the positioning functionality of the nRF9160 including GNSS and cellular positioning.
It can be used to retrieve the location of the device in the form of events containing a position, velocity and time (PVT) structure.

> 

Reviewer 3: ```suggestion
```

Features
********

This section documents the various features implemented by the module.

Location control
================

The module uses the :ref:`lib_location` to communicate with the nRF9160 modem and control its GNSS and LTE neighbor cell measurement functionalities.
A location request is started when the module receives an ``APP_EVT_DATA_GET`` event and the ``APP_DATA_LOCATION`` type is listed in the event's ``data_list`` member containing the data types that shall be sampled.

> The default configuration for the location request is set as follows:

Author: If you use active voice, who is the one setting the default configuration?
In reality, it's the location module that we are documenting here.
With active voice, for me it sounds like it would maybe be user setting the default configuration but I'll let you judge that.

> * GNSS set as the first priority method and cellular positioning as the second priority.

Author: I'll close this and we'll resolve this topic in comment https://github.com/nrfconnect/sdk-nrf/pull/8631#discussion_r1010415322

  Used location methods can be configured at build time
  (:ref:`CONFIG_DATA_SAMPLE_GNSS_DEFAULT <CONFIG_DATA_SAMPLE_GNSS_DEFAULT>` and
  :ref:`CONFIG_DATA_SAMPLE_NEIGHBOR_CELLS_DEFAULT <CONFIG_DATA_SAMPLE_NEIGHBOR_CELLS_DEFAULT>`)

>   or to the dynamically updated value set by the cloud side and stored to flash.

Reviewer 3: **suggestion:** consider clarifying "the dynamically updated value".

Author: This part is direct copy-paste from earlier GNSS module documentation. But maybe we could actually say:
```
Methods that should not be used can be configured by adding them into ``No Data List``. See :ref:`Real-time configurations <real_time_configs>` for more information.
```
What do you think @coderbyheart?

Reviewer 3: Yes, sounds good!

* Timeout set to either the value configured at build time
  (:ref:`CONFIG_DATA_LOCATION_TIMEOUT_SECONDS <CONFIG_DATA_LOCATION_TIMEOUT_SECONDS>`)

>   or to the dynamically updated value set by the cloud side and stored to flash.

Reviewer 3: **suggestion:** consider clarifying "the dynamically updated value".

Author: I've modified this to be aligned with the previous bullet point:
`Timeout which can be configured as part of :ref:`Real-time configurations <real_time_configs>``

  Values stored in flash take precedence.

The location module receives configuration updates as payload in the following two events:

* ``DATA_EVT_CONFIG_INIT`` - The event contains the value configured at build time or received from cloud and stored to flash.
  The event is distributed from the data module as part of the application initialization.
* ``DATA_EVT_CONFIG_READY`` - The event is received when the data module has processed an incoming message from cloud with a configuration update.

When a position has been retrieved, the ``LOCATION_MODULE_EVT_DATA_READY`` event is sent from the module.

> If location request failed, ``LOCATION_MODULE_EVT_DATA_NOT_READY`` is sent.

Reviewer 3: ```suggestion
If the location request failed, ``LOCATION_MODULE_EVT_DATA_NOT_READY`` is sent.
```

The application uses the :ref:`lib_modem_antenna` library for configuring the LNA.
The library has Kconfig options to control the LNA using either the COEX0 or MAGPIO interface of the nRF9160.
See the library documentation for more details on how to configure the antenna and LNA.

GPS assistance data
===================

The location module receives requests for GPS assistance data from the :ref:`lib_location` library.
When an A-GPS request is received, the module distributes it to the other modules as a ``LOCATION_MODULE_EVT_AGPS_NEEDED`` event that contains information about the type of assistance data needed.
Providing the requested A-GPS data may reduce the time it takes to acquire a GNSS fix.

Module internals
================

The location module is a threadless module in terms of message processing.
In this sense, it differs from the other modules.
It has a thread for processing location events from the :ref:`lib_location` library, whereas other modules use a thread for processing messages.
The thread is used to offload data processing from the location callbacks coming from the :ref:`lib_location` library, and no time consuming tasks should be performed there.

> All incoming events from other modules are handled in the context of the Application Event Manager callback, as they all complete fast enough to not require a separate thread.

Reviewer 3: ```suggestion
All incoming events from other modules are handled in the context of the Application Event Manager callback, because they all complete fast enough to not require a separate thread.
```

> .. _configuration_options:

Author: I cannot find this link being used so removing for now and seeing if it works.

Configuration options
*********************

.. _CONFIG_LOCATION_MODULE:

CONFIG_LOCATION_MODULE

>    Enables the location module. The module can report data either as a structure of position, velocity, and time (PVT) information, or as an NMEA string of GGA type.

Author: Yep, removed most of the description which is aligned with some modules. And some doesn't even have the module's on enable config mentioned.
